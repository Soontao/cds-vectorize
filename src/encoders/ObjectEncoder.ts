/* eslint-disable camelcase */
import { inspect } from "util";
import BoolEncoder from "./BoolEncoder.js";
import CategoryEncoder from "./CategoryEncoder.js";
import DateTimeEncoder from "./DateTimeEncoder.js";
import Encoder from "./Encoder.js";
import FixedListEncoder from "./FixedListEncoder.js";
import ObjectMetadata, { metadataValidator } from "./Metadata.js";
import NumericEncoder from "./NumericEncoder.js";
import StatisticListEncoder from "./StatisticListEncoder.js";
import UUIDEncoder from "./UUIDEncoder.js";
import { Vector } from "./type.js";

export function sort(meta: ObjectMetadata): ObjectMetadata {
  if (meta._sorted) {
    return meta;
  }
  const newMeta = { ...meta };
  newMeta.properties = meta.properties
    .map((v) => {
      if (v.type == "object") {
        const newV = { ...v };
        newV.meta = sort(v.meta as any);
        return newV;
      }
      if (v.type === "category" && v.values) {
        v.values = v.values!.sort();
      }
      return v;
    })
    .sort((p1, p2) => (p1.name > p2.name ? 1 : -1));
  Object.defineProperty(newMeta, "_sorted", { value: true, enumerable: false });
  return newMeta;
}

export function fillEncoders(meta: ObjectMetadata) {
  if (meta._encoder_filled) {
    return meta;
  }
  for (const property of meta.properties) {
    if (property._encoder !== undefined) {
      continue;
    }
    if (property.meta) {
      sortMetaAndFillEncoders(property.meta!); // inner meta create encoders firstly
    }
    switch (property.type) {
      case "bool":
        property._encoder = new BoolEncoder();
        break;
      case "category":
        property._encoder = new CategoryEncoder(property.values as Array<any>);
        break;
      case "datetime":
        property._encoder = new DateTimeEncoder();
        break;
      case "numeric":
        property._encoder = new NumericEncoder();
        break;
      case "object":
        property._encoder = new ObjectEncoder(property.meta!);
        break;
      case "uuid":
        property._encoder = new UUIDEncoder();
        break;
      case "fixed_object_list":
        property._encoder = new FixedListEncoder(property.meta!, property.position_dict!);
        break;
      case "statistic_object_list":
        property._encoder = new StatisticListEncoder(property.meta!);
        break;
      default:
        throw new TypeError(`cannot handle type ${property.type} for ${property.name}`);
    }
  }
  calculateObjectVecLength(meta);
  Object.defineProperty(meta, "_encoder_filled", { value: true, enumerable: false });
  return meta;
}

export function sortMetaAndFillEncoders(meta: ObjectMetadata): ObjectMetadata {
  if (meta._valid == undefined) {
    const isValid = metadataValidator(meta as any);
    if (!isValid) {
      throw new Error(`not valid metadata, please check ${inspect(metadataValidator.errors)}`);
    }
  }
  Object.defineProperty(meta, "_valid", { value: true, enumerable: false });
  return fillEncoders(sort(meta));
}

export function calculateObjectVecLength(meta: ObjectMetadata): number {
  if (meta._length) {
    return meta._length;
  }
  const _length = meta.properties.reduce((acc, property) => acc + property._encoder!.length, 0);
  Object.defineProperty(meta, "_length", { value: _length, enumerable: false });
  return _length;
}

/**
 * @ai
 * @human
 */
export class ObjectEncoder<T> implements Encoder<T> {
  #meta: ObjectMetadata;

  constructor(meta: ObjectMetadata) {
    this.#meta = sortMetaAndFillEncoders(meta);
  }

  features(name: string = "root"): string[] {
    return this.#meta.properties.map((p) => p._encoder!.features(p.name).map((pName) => `${name}_${pName}`)).flat();
  }

  encode(value: T): Vector {
    if (typeof value !== "object" || value === null) {
      throw new Error("Invalid value. Expected an object.");
    }

    const encodedVector: number[] = [];

    for (const property of this.#meta.properties) {
      encodedVector.push(...property._encoder!.encode((value as any)[property.name]));
    }

    return encodedVector;
  }

  decode(vec: Vector): T {
    if (vec.length !== this.length) {
      throw new Error("Invalid vector length");
    }

    const decodedObject: any = {};
    let index = 0;

    for (const property of this.#meta.properties) {
      const encoder = property._encoder!;
      const value = encoder.decode(vec.slice(index, index + encoder.length));
      decodedObject[property.name] = value;
      index += property._encoder!.length;
    }

    return decodedObject as T;
  }

  get length(): number {
    return this.#meta._length!;
  }
}

export default ObjectEncoder;
