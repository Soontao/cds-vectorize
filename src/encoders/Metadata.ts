/* eslint-disable camelcase */
/* eslint-disable @typescript-eslint/indent */
import { Schema, validator } from "@exodus/schemasafe";
import BoolEncoder from "./BoolEncoder.js";
import CategoryEncoder from "./CategoryEncoder.js";
import DateTimeEncoder from "./DateTimeEncoder.js";
import Encoder from "./Encoder.js";
import FixedListEncoder from "./FixedListEncoder.js";
import MultiCategoryEncoder from "./MultiCategoryEncoder.js";
import MurmurEncoder from "./MurmurEncoder.js";
import NumericEncoder from "./NumericEncoder.js";
import ObjectEncoder from "./ObjectEncoder.js";
import StatisticListEncoder from "./StatisticListEncoder.js";
import UUIDEncoder from "./UUIDEncoder.js";

export interface Property {
  /**
   * name of property
   */
  name: string;
  /**
   * type
   */
  type:
    | "category"
    | "murmur_hash"
    | "multi_category"
    | "bool"
    | "uuid"
    | "numeric"
    | "datetime"
    | "object"
    | "fixed_object_list"
    | "statistic_object_list";

  /**
   * possible values for category
   */
  values?: Array<any>;

  /**
   * metadata for object
   */
  meta?: ObjectMetadata;

  /**
   * only for fixed_object_list, used to match input list item, then assign to correct position
   */
  position_dict?: Array<any>;

  /**
   * only for hash encoders like murmur_hash
   */
  hash_seed?: number;

  /**
   * internal _encoder
   */
  _encoder?: Encoder;
}

export interface ObjectMetadata {
  properties: Array<Property>;
  readonly _encoder_filled?: boolean;
  readonly _sorted?: boolean;
  readonly _length?: number;
  readonly _valid?: boolean;
}

const metadataSchema: Schema = {
  $schema: "http://json-schema.org/draft-07/schema#",
  definitions: {
    Property: {
      type: "object",
      properties: {
        name: { type: "string" },
        type: {
          type: "string",
          enum: [
            "category",
            "murmur_hash",
            "multi_category",
            "bool",
            "uuid",
            "numeric",
            "datetime",
            "object",
            "fixed_object_list",
            "statistic_object_list",
          ],
        },
        values: {
          type: "array",
          minItems: 1,
          items: { type: ["string", "number", "boolean", "null"] },
        },
        meta: { $ref: "#/definitions/ObjectMetadata" },
        position_dict: { type: "array" },
        hash_seed: { type: "number" },
      },
      required: ["name", "type"],
      if: {
        properties: { type: { const: ["category", "multi_category"] } },
      },
      then: {
        required: ["values"],
      },
      else: {
        if: {
          properties: {
            type: { const: "object" },
          },
        },
        then: {
          required: ["meta"],
        },
        else: {
          if: {
            properties: {
              type: { const: "fixed_object_list" },
            },
          },
          then: {
            required: ["position_dict", "meta"],
          },
        },
      },
    },
    ObjectMetadata: {
      type: "object",
      properties: {
        properties: {
          type: "array",
          minItems: 1,
          items: { $ref: "#/definitions/Property" },
        },
      },
      required: ["properties"],
    },
  },
  type: "object",
  properties: {
    properties: {
      type: "array",
      minItems: 1,
      items: { $ref: "#/definitions/Property" },
    },
  },
  required: ["properties"],
};

export function mapEncoder(property: Property) {
  switch (property.type) {
    case "bool":
      return new BoolEncoder(property);
    case "category":
      return new CategoryEncoder(property);
    case "multi_category":
      return new MultiCategoryEncoder(property);
    case "datetime":
      return new DateTimeEncoder(property);
    case "numeric":
      return new NumericEncoder(property);
    case "object":
      return new ObjectEncoder(property);
    case "uuid":
      return new UUIDEncoder(property);
    case "fixed_object_list":
      return new FixedListEncoder(property);
    case "statistic_object_list":
      return new StatisticListEncoder(property);
    case "murmur_hash":
      return new MurmurEncoder(property);
    default:
      throw new TypeError(`cannot handle type ${property.type} for ${property.name}`);
  }
}

export function createObjectMetadata(objectMetadata: ObjectMetadata) {
  return objectMetadata;
}

export const metadataValidator = validator(metadataSchema, { includeErrors: true });

export default ObjectMetadata;
