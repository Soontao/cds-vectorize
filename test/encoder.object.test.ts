// @ai
// @human
import { ObjectMetadata } from "../src/encoders/Metadata.js";
import { ObjectEncoder, sort, sortMetaAndFillEncoders } from "../src/encoders/ObjectEncoder.js";

describe("ObjectEncoder Test Suite", () => {
  it("should sort properties alphabetically", () => {
    // Arrange
    const unsortedMeta: ObjectMetadata = {
      properties: [
        { name: "zeta", type: "category", values: ["A", "B"] },
        { name: "alpha", type: "numeric" },
        { name: "beta", type: "bool" },
      ],
    };

    // Act
    const sortedMeta = sort(unsortedMeta);

    // Assert
    const expectedSortedMeta: ObjectMetadata = {
      properties: [
        { name: "alpha", type: "numeric" },
        { name: "beta", type: "bool" },
        { name: "zeta", type: "category", values: ["A", "B"] },
      ],
    };

    expect(sortedMeta).toMatchObject(expectedSortedMeta);
  });

  it("should recursively sort properties within objects", () => {
    // Arrange
    const unsortedMeta: ObjectMetadata = {
      properties: [
        { name: "zeta", type: "category", values: ["A", "B"] },
        {
          name: "nested",
          type: "object",
          meta: {
            properties: [
              { name: "gamma", type: "numeric" },
              { name: "alpha", type: "bool" },
            ],
          },
        },
        { name: "alpha", type: "numeric" },
      ],
    };

    // Act
    const sortedMeta = sort(unsortedMeta);

    // Assert
    const expectedSortedMeta: ObjectMetadata = {
      properties: [
        { name: "alpha", type: "numeric" },
        {
          name: "nested",
          type: "object",
          meta: {
            properties: [
              { name: "alpha", type: "bool" },
              { name: "gamma", type: "numeric" },
            ],
          },
        },
        { name: "zeta", type: "category", values: ["A", "B"] },
      ],
    };

    expect(sortedMeta).toMatchObject(expectedSortedMeta);
  });

  it("should sort category values alphabetically", () => {
    // Arrange
    const unsortedMeta: ObjectMetadata = {
      properties: [
        { name: "alpha", type: "category", values: ["C", "B", "A"] },
        { name: "beta", type: "category", values: ["Z", "X", "Y"] },
        { name: "gamma", type: "numeric" },
      ],
    };

    // Act
    const sortedMeta = sort(unsortedMeta);

    // Assert
    const expectedSortedMeta: ObjectMetadata = {
      properties: [
        { name: "alpha", type: "category", values: ["A", "B", "C"] },
        { name: "beta", type: "category", values: ["X", "Y", "Z"] },
        { name: "gamma", type: "numeric" },
      ],
    };

    expect(sortedMeta).toMatchObject(expectedSortedMeta);
  });

  it("should encode and decode an object correctly", () => {
    // Sample metadata for testing
    const metadata: any = {
      properties: [
        { name: "prop_bool", type: "bool" },
        { name: "prop_category", type: "category", values: ["A", "B", "C"] },
        { name: "prop_datetime", type: "datetime" },
        { name: "prop_numeric", type: "numeric" },
        {
          name: "prop_object",
          type: "object",
          meta: {
            properties: [
              { name: "nested_bool", type: "bool" },
              { name: "nested_category", type: "category", values: ["X", "Y", "Z"] },
            ],
          },
        },
      ],
    };

    // Sample object data for testing
    const testData = {
      prop_bool: true,
      prop_category: "B",
      prop_datetime: "2023-01-01T12:30:00.000Z",
      prop_numeric: 42,
      prop_object: {
        nested_bool: false,
        nested_category: "Z",
      },
    };

    // Create an instance of ObjectEncoder
    const objectEncoder = new ObjectEncoder(metadata);

    expect(objectEncoder.length).toBe(18);

    // Encode the test data
    const encodedVector = objectEncoder.encode(testData);

    expect(encodedVector.length).toBe(objectEncoder.length);

    expect(encodedVector).toMatchSnapshot();

    // Decode the encoded vector
    const decodedObject = objectEncoder.decode(encodedVector);

    expect(decodedObject).toMatchSnapshot();

    // Assertions
    expect(decodedObject).toEqual(testData);
  });

  // @ai
  it("should encode and decode correctly with nested object and list object", () => {
    // Sample metadata for testing
    const metadata: any = {
      properties: [
        { name: "prop_bool", type: "bool" },
        { name: "prop_category", type: "category", values: ["A", "B", "C"] },
        { name: "prop_datetime", type: "datetime" },
        { name: "prop_numeric", type: "numeric" },
        {
          name: "prop_object",
          type: "object",
          meta: {
            properties: [
              { name: "nested_bool", type: "bool" },
              { name: "nested_category", type: "category", values: ["X", "Y", "Z"] },
              {
                name: "nested_list",
                type: "fixed_object_list",
                meta: {
                  properties: [
                    { name: "item_name", type: "category", values: ["Item1", "Item2", "Item3"] },
                    { name: "item_value", type: "numeric" },
                  ],
                },
                position_dict: [{ item_name: "Item1" }, { item_name: "Item2" }],
              },
            ],
          },
        },
      ],
    };

    // Sample object data for testing
    const testData = {
      prop_bool: true,
      prop_category: "B",
      prop_datetime: "2023-01-01T12:30:00.000Z",
      prop_numeric: 42,
      prop_object: {
        nested_bool: false,
        nested_category: "Z",
        nested_list: [
          { item_name: "Item1", item_value: 10 },
          { item_name: "Item2", item_value: 20 },
        ],
      },
    };

    // Create an instance of ObjectEncoder
    const objectEncoder = new ObjectEncoder(metadata);
    const features = objectEncoder.features();

    expect(features).toMatchSnapshot();

    // Encode the test data
    const encodedVector = objectEncoder.encode(testData);

    expect(encodedVector).toHaveLength(features.length);
    expect(encodedVector).toMatchSnapshot();

    // Decode the encoded vector
    const decodedObject = objectEncoder.decode(encodedVector);

    // Assertions
    expect(decodedObject).toEqual(testData);
  });

  it("should encode and decode correctly with nested object and list object", () => {
    // Sample metadata for testing
    const metadata: ObjectMetadata = {
      properties: [
        { name: "object_code", type: "murmur_hash" },
        { name: "prop_bool", type: "bool" },
        { name: "prop_category", type: "category", values: ["A", "B", "C"] },
        { name: "prop_datetime", type: "datetime" },
        { name: "prop_numeric", type: "numeric" },
        {
          name: "prop_object",
          type: "object",
          meta: {
            properties: [
              { name: "nested_bool", type: "bool" },
              { name: "nested_category", type: "category", values: ["X", "Y", "Z"] },
              {
                name: "nested_list",
                type: "statistic_object_list", // Use the new type
                meta: {
                  properties: [
                    { name: "item_name", type: "category", values: ["Item1", "Item2", "Item3"] },
                    { name: "item_value", type: "numeric" },
                  ],
                },
                position_dict: [{ item_name: "Item1" }, { item_name: "Item2" }],
              },
            ],
          },
        },
      ],
    };

    // Sample object data for testing
    const testData = {
      object_code: "T_001",
      prop_bool: true,
      prop_category: "B",
      prop_datetime: "2023-01-01T12:30:00.000Z",
      prop_numeric: 42,
      prop_object: {
        nested_bool: false,
        nested_category: "Z",
        nested_list: [
          { item_name: "Item1", item_value: 10 },
          { item_name: "Item2", item_value: 20 },
        ],
      },
    };

    // Create an instance of ObjectEncoder
    const objectEncoder = new ObjectEncoder(metadata);
    const features = objectEncoder.features();

    expect(features).toMatchSnapshot();

    // Encode the test data
    const encodedVector = objectEncoder.encode(testData);

    expect(encodedVector).toHaveLength(features.length);
    expect(encodedVector).toMatchSnapshot();
  });

  it("Throws error for invalid ObjectMetadata", () => {
    const invalidObjectMetadata = {
      // Missing required properties field
      _encoder_filled: true,
      _sorted: false,
      _length: 5,
    };

    expect(() => sortMetaAndFillEncoders(invalidObjectMetadata as any)).toThrowErrorMatchingSnapshot();
  });
});
