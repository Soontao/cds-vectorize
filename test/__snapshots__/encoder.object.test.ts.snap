// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ObjectEncoder Test Suite Throws error for invalid ObjectMetadata 1`] = `"not valid metadata, please check [ { keywordLocation: '#/required', instanceLocation: '#/properties' } ]"`;

exports[`ObjectEncoder Test Suite should encode and decode an object correctly 1`] = `
[
  1,
  0,
  0,
  1,
  0,
  2023,
  1,
  1,
  52,
  12,
  30,
  0,
  42,
  0,
  1,
  0,
  0,
  1,
]
`;

exports[`ObjectEncoder Test Suite should encode and decode an object correctly 2`] = `
{
  "prop_bool": true,
  "prop_category": "B",
  "prop_datetime": "2023-01-01T12:30:00.000Z",
  "prop_numeric": 42,
  "prop_object": {
    "nested_bool": false,
    "nested_category": "Z",
  },
}
`;

exports[`ObjectEncoder Test Suite should encode and decode correctly with nested object and list object 1`] = `
[
  "root_prop_bool_is_true",
  "root_prop_bool_is_false",
  "root_prop_category_is_A",
  "root_prop_category_is_B",
  "root_prop_category_is_C",
  "root_prop_datetime_year",
  "root_prop_datetime_month",
  "root_prop_datetime_day",
  "root_prop_datetime_isoWeek",
  "root_prop_datetime_hour",
  "root_prop_datetime_minute",
  "root_prop_datetime_second",
  "root_prop_numeric",
  "root_prop_object_nested_bool_is_true",
  "root_prop_object_nested_bool_is_false",
  "root_prop_object_nested_category_is_X",
  "root_prop_object_nested_category_is_Y",
  "root_prop_object_nested_category_is_Z",
  "root_prop_object_nested_list_0_item_name_is_Item1",
  "root_prop_object_nested_list_0_item_name_is_Item2",
  "root_prop_object_nested_list_0_item_name_is_Item3",
  "root_prop_object_nested_list_0_item_value",
  "root_prop_object_nested_list_1_item_name_is_Item1",
  "root_prop_object_nested_list_1_item_name_is_Item2",
  "root_prop_object_nested_list_1_item_name_is_Item3",
  "root_prop_object_nested_list_1_item_value",
]
`;

exports[`ObjectEncoder Test Suite should encode and decode correctly with nested object and list object 2`] = `
[
  1,
  0,
  0,
  1,
  0,
  2023,
  1,
  1,
  52,
  12,
  30,
  0,
  42,
  0,
  1,
  0,
  0,
  1,
  1,
  0,
  0,
  10,
  0,
  1,
  0,
  20,
]
`;

exports[`ObjectEncoder Test Suite should encode and decode correctly with nested object and list object 3`] = `
[
  "root_prop_bool_is_true",
  "root_prop_bool_is_false",
  "root_prop_category_is_A",
  "root_prop_category_is_B",
  "root_prop_category_is_C",
  "root_prop_datetime_year",
  "root_prop_datetime_month",
  "root_prop_datetime_day",
  "root_prop_datetime_isoWeek",
  "root_prop_datetime_hour",
  "root_prop_datetime_minute",
  "root_prop_datetime_second",
  "root_prop_numeric",
  "root_prop_object_nested_bool_is_true",
  "root_prop_object_nested_bool_is_false",
  "root_prop_object_nested_category_is_X",
  "root_prop_object_nested_category_is_Y",
  "root_prop_object_nested_category_is_Z",
  "root_prop_object_nested_list_item_name_is_Item1_max",
  "root_prop_object_nested_list_item_name_is_Item2_max",
  "root_prop_object_nested_list_item_name_is_Item3_max",
  "root_prop_object_nested_list_item_value_max",
  "root_prop_object_nested_list_item_name_is_Item1_min",
  "root_prop_object_nested_list_item_name_is_Item2_min",
  "root_prop_object_nested_list_item_name_is_Item3_min",
  "root_prop_object_nested_list_item_value_min",
  "root_prop_object_nested_list_item_name_is_Item1_avg",
  "root_prop_object_nested_list_item_name_is_Item2_avg",
  "root_prop_object_nested_list_item_name_is_Item3_avg",
  "root_prop_object_nested_list_item_value_avg",
  "root_prop_object_nested_list_item_name_is_Item1_diff",
  "root_prop_object_nested_list_item_name_is_Item2_diff",
  "root_prop_object_nested_list_item_name_is_Item3_diff",
  "root_prop_object_nested_list_item_value_diff",
  "root_prop_object_nested_list_item_name_is_Item1_variance",
  "root_prop_object_nested_list_item_name_is_Item2_variance",
  "root_prop_object_nested_list_item_name_is_Item3_variance",
  "root_prop_object_nested_list_item_value_variance",
  "root_prop_object_nested_list_item_name_is_Item1_sum",
  "root_prop_object_nested_list_item_name_is_Item2_sum",
  "root_prop_object_nested_list_item_name_is_Item3_sum",
  "root_prop_object_nested_list_item_value_sum",
]
`;

exports[`ObjectEncoder Test Suite should encode and decode correctly with nested object and list object 4`] = `
[
  1,
  0,
  0,
  1,
  0,
  2023,
  1,
  1,
  52,
  12,
  30,
  0,
  42,
  0,
  1,
  0,
  0,
  1,
  1,
  1,
  0,
  20,
  0,
  0,
  0,
  10,
  0.5,
  0.5,
  0,
  15,
  1,
  1,
  0,
  10,
  0.25,
  0.25,
  0,
  25,
  1,
  1,
  0,
  30,
]
`;
